ext/dom/document.c

```diff
@@ -1724,23 +1726,23 @@ static void _dom_document_relaxNG_validate(INTERNAL_FUNCTION_PARAMETERS, int typ
 		RETURN_THROWS();
 	}
 
-	if (source_len == 0) {
-		php_error_docref(NULL, E_WARNING, "Invalid Schema source");
-		RETURN_FALSE;
+	if (!source_len) {
+		zend_argument_value_error(1, "must not be empty");
+		RETURN_THROWS();
 	}
 
 	DOM_GET_OBJ(docp, id, xmlDocPtr, intern);
 
 	switch (type) {
 	case DOM_LOAD_FILE:
 		if (CHECK_NULL_PATH(source, source_len)) {
-			php_error_docref(NULL, E_WARNING, "Invalid RelaxNG file source");
-			RETURN_FALSE;
+			zend_argument_value_error(1, "must not contain any null bytes");
+			RETURN_THROWS();
 		}
 		valid_file = _dom_get_valid_file_path(source, resolved_path, MAXPATHLEN);
 		if (!valid_file) {
-			php_error_docref(NULL, E_WARNING, "Invalid RelaxNG file source");
-			RETURN_FALSE;
+			zend_argument_value_error(1, "is not a valid path");
+			RETURN_THROWS();
```

Comment:
Do I understand right that this triggers for too long paths? In that case, we decided that this case should not throw, and rather be treated at the same error level as a non-existing file (grep "exceeds the maximum allowed length").

Reply:
Also, I'm not really clear on what other things this could be checking. Path length is one of them, but going down the expand_filepath and virtual_file_ex rabbit hole, there's probably more to it than that.

Reply:
The `_dom_get_valid_file_path()` method also appears to be doing platform specific checks for invalid paths, such as Windows systems that do not allow any `*` or `?` in paths. Hence the generic "is not a valid path" message.

My rationale for making this throw is that it is not fundamentally different from throwing on null bytes. I generalized it to "nonsensical path names must throw".

I could add a separate check for long paths and make those warnings. It does feel like a strange special case to me though. The question is: why is a path containing illegal characters more nonsensical than a path that is too long?

Reply:
> I could add a separate check for long paths and make those warnings. It does feel like a strange special case to me though. The question is: why is a path containing illegal characters more nonsensical than a path that is too long?

It isn't. Neither a too long path, nor invalid characters in the path should throw. These are environment error conditions and do not necessarily indicate programmer error. (Stepping away from DOM: If the user provides a path on the CLI, the programmer should *not* be forced to validate it against some platform-specific requirements that are probably not even documented anywhere before passing it to fopen, in order to avoid an exception in some edge case.)

Reply:
I think the only argument that can be made here is that null bytes shouldn't be throwing either, which is certainly a possibility. Null bytes are somewhat special in that they indicate active malicious intent, and in that they are implicitly excluded from certain input sources. E.g. if you accept a file name as a command line argument, it cannot contain null bytes in the first place.

Reply:
Ok. The RFC mainly covers promotion of programming errors, which has a rather large gray area surrounding it. I do agree that null bytes are unlikely to originate from programming errors. I will adjust the PR to make these generate a warning in stead. Likewise for long or otherwise invalid paths.

Reply:
One issue regarding null bytes: As @Girgias pointed out, the `p` and `P` argument parsing formats also throw on null bytes. I'm not sure for how long this has been the case and what to do with it now.